File: Pinball App 2/Pinball App 2/practice/ScoreScannerReadingSupport.swift
import CoreGraphics

func scoreScannerLockedReading(from snapshot: ScoreStabilityService.Snapshot?) -> ScoreScannerLockedReading? {
    guard let snapshot, let reading = snapshot.dominantReading else { return nil }
    return ScoreScannerLockedReading(
        score: reading.score,
        formattedScore: reading.formattedScore,
        rawText: reading.rawText
    )
}

func scoreScannerDisplayedReading(
    from snapshot: ScoreStabilityService.Snapshot,
    bestCandidate: ScoreScannerCandidate?
) -> ScoreScannerLockedReading? {
    if let locked = scoreScannerLockedReading(from: snapshot) {
        return locked
    }
    guard let bestCandidate else { return nil }
    return scoreScannerReading(from: bestCandidate)
}

func scoreScannerReading(from candidate: ScoreScannerCandidate) -> ScoreScannerLockedReading {
    ScoreScannerLockedReading(
        score: candidate.normalizedScore,
        formattedScore: candidate.formattedScore,
        rawText: candidate.rawText
    )
}

func scoreScannerFilteredAnalysis(
    _ analysis: ScoreOCRAnalysis,
    minimumDigitCount: Int,
    minimumHorizontalPadding: CGFloat,
    minimumVerticalPadding: CGFloat
) -> ScoreOCRAnalysis {
    let filteredCandidates = analysis.candidates.filter { candidate in
        candidate.digitCount >= minimumDigitCount &&
        candidate.boundingBox.minX >= minimumHorizontalPadding &&
        candidate.boundingBox.maxX <= (1 - minimumHorizontalPadding) &&
        candidate.boundingBox.minY >= minimumVerticalPadding &&
        candidate.boundingBox.maxY <= (1 - minimumVerticalPadding)
    }
    return ScoreOCRAnalysis(bestCandidate: filteredCandidates.first, candidates: filteredCandidates)
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/ScoreScannerReadingSupport.kt
package com.pillyliu.pinprofandroid.practice

internal fun scoreScannerFilteredAnalysis(
    analysis: ScoreScannerAnalysis,
    minimumDigitCount: Int,
    minimumHorizontalPadding: Float,
    minimumVerticalPadding: Float,
): ScoreScannerAnalysis {
    val filteredCandidates = analysis.candidates.filter { candidate ->
        candidate.digitCount >= minimumDigitCount &&
            candidate.boundingBox.left >= minimumHorizontalPadding &&
            candidate.boundingBox.right <= (1f - minimumHorizontalPadding) &&
            candidate.boundingBox.top >= minimumVerticalPadding &&
            candidate.boundingBox.bottom <= (1f - minimumVerticalPadding)
    }

    return ScoreScannerAnalysis(
        bestCandidate = filteredCandidates.firstOrNull(),
        candidates = filteredCandidates,
    )
}

internal fun scoreScannerLockedReading(
    snapshot: ScoreScannerStabilityService.Snapshot?,
): ScoreScannerLockedReading? {
    val reading = snapshot?.dominantReading ?: return null
    return ScoreScannerLockedReading(
        score = reading.score,
        formattedScore = reading.formattedScore,
        rawText = reading.rawText,
        confidence = reading.confidence,
        averageConfidence = snapshot.averageConfidence,
    )
}

internal fun scoreScannerDisplayedReading(
    snapshot: ScoreScannerStabilityService.Snapshot?,
    bestCandidate: ScoreScannerCandidate?,
): ScoreScannerLockedReading? {
    scoreScannerLockedReading(snapshot)?.let { return it }
    val candidate = bestCandidate ?: return null
    return scoreScannerReadingFrom(candidate)
}

internal fun scoreScannerReadingFrom(candidate: ScoreScannerCandidate): ScoreScannerLockedReading {
    return ScoreScannerLockedReading(
        score = candidate.normalizedScore,
        formattedScore = candidate.formattedScore,
        rawText = candidate.rawText,
        confidence = candidate.confidence,
        averageConfidence = candidate.confidence,
    )
}

internal fun shouldAttemptScoreScannerLiveBitmapFallback(
    now: Long,
    lastFallbackAt: Long,
    snapshot: ScoreScannerStabilityService.Snapshot?,
    sourceBestCandidate: ScoreScannerCandidate?,
    liveBitmapFallbackIntervalMs: Long,
    strongLiveCandidateDigitCount: Int,
    strongLiveCandidateFormatQuality: Int,
): Boolean {
    if (now - lastFallbackAt < liveBitmapFallbackIntervalMs) return false
    if (snapshot?.state == ScoreScannerStatus.Locked) return false
    val candidate = sourceBestCandidate ?: return true
    return candidate.digitCount < strongLiveCandidateDigitCount ||
        candidate.formatQuality < strongLiveCandidateFormatQuality
}

internal fun mergeScoreScannerAnalyses(
    primary: ScoreScannerAnalysis,
    secondary: ScoreScannerAnalysis,
): ScoreScannerAnalysis {
    val rankedCandidates = ScoreScannerParsingService.rankCandidates(
        primary.candidates + secondary.candidates
    )
    return ScoreScannerAnalysis(
        bestCandidate = rankedCandidates.firstOrNull(),
        candidates = rankedCandidates,
    )
}
